### PR TITLE
Handling of indentation for modules with a `with`

### LINKED ISSUE #248: Handling of indentation for modules with a `with`
As of f97a163, the following
```ocaml
module MFloat : SERIALISABLE
  with type t = float =
struct
  type t = float [@@deriving yojson]
  let _key = "float"
end
```
gets formatted as
```ocaml
module MFloat: SERIALISABLE with
  type t = float = struct
  type t = float [@@deriving yojson]
  let _key = "float"
end
```
I think this yields the same considerations as #219 in term of readability: the `struct` gets lost, and the `type t = float` from the `with` statement can get confused for a declaration in the module.

In a similar vain, I am not very convinced by 
```ocaml
module MOption
  (A: SERIALISABLE)
  : SERIALISABLE = struct
  type t = A.t option [@@deriving yojson]
  let _key = A._key ^ "-option"
end
```
with no visual clue about when module argument/types end and when the module definition starts.

### PR DESCRIPTION
**Pull request by Xophmeister**
* Resolves #248 
* [x] Based on #282 (merge first to avoid conflicts)

### CODE REVIEW

**Comment by Xophmeister** [comment]
This is a work in progress.

I have added the input softline between the `=` and a structure, in a module definition. However, somewhere between f97a163...f7d1004, a rule was introduced that breaks the indentation. (Amongst some more subtle differences.).

Expected:
```ocaml
module MFloat : SERIALISABLE
  with type t = float =
struct
  type t = float [@@deriving yojson]
  let _key = "float"
end
```

Actual (as of ad0b2b6):
```ocaml
module MFloat: SERIALISABLE with
    type t = float =
struct
  type t = float [@@deriving yojson]
  let _key = "float"
end
```

An extra pair of eyes would be appreciated on this one :eyes:

**Comment by Xophmeister** [comment]
The indentation bug was introduced in 12bf707e78e4, with this rule:

https://github.com/tweag/topiary/blob/12bf707e78e47497f1269d4d96a77477df2315fe/languages/ocaml.scm#L1290-L1303

**Comment by Xophmeister** [comment]
@nbacquey I can fix the indentation if I add a `@do_nothing` rule against any `with` anonymous node that appears:

```scm
(module_binding
  (#scope_id! "module_binding_before_equal")
  (module_name) @append_indent_start @begin_scope
  "="? @do_nothing
  [
    (signature)? @do_nothing
    [
      (functor_type)
      (module_type_constraint
        "with"? @do_nothing
      )
    ] @append_indent_end @end_scope
  ]
  .
)
```

However, that then breaks the indentation on other examples with a `with`:
```diff
 module MUnit: SERIALISABLE with
-  type t = unit
+type t = unit

     (* ...snip... *)
     module Protocol_implementation:
-      PVM.S with
-      type context = Context.t
-      and type state = Context.tree
-      and type proof = Context.Proof.tree Context.Proof.t
+    PVM.S with
+    type context = Context.t
+    and type state = Context.tree
+    and type proof = Context.Proof.tree Context.Proof.t
```

I _think_, therefore, that there was nothing wrong with this rule, as such, but the double indentation is getting applied elsewhere. However, the tree for the `type t = float` is:

* `(constrain_type)`
  * Anonymous `type`
  * `(type_constructor_path)`
    * `(type_constructor)`
  * Anonymous `=`
  * `(type_constructor_path)`
    * `(type_constructor)`

I don't see any rule that would match any part of this subtree... Ideas?

**Comment by nbacquey** [comment]
The problem is indeed that the rule you identified applies, where it should not (it's designed to apply only for module bindings without an `=` sign, but it obviously fails at that. I'm thinking about a way to fix it.

On another note, it allowed me to identify yet another problem with tree-sitter query grammar, which I'll be documenting

**Comment by nbacquey** [comment]
This new version of the rule fixes the problem:
```scheme
(module_binding
  (#scope_id! "module_binding_before_equal")
  (module_name) @append_indent_start @begin_scope
  [
    (functor_type)
    (module_type_constraint)
  ] @append_indent_end @end_scope
  "="? @do_nothing
)
```
I removed the `(signature)` part, because it did nothing to the tests. The real modification comes from pushing the `=` at the end.

Also, instead of adding a new unopinionated rule to keep input line breaks, I would prefer to be opinionated and add a scoped line break using the query just above the one I modified. Do you want to handle it @Xophmeister ? Otherwise I can draft something.

File: languages/ocaml.scm [hunk 1]
```diff
@@ -1386,22 +1386,20 @@
...
   (#scope_id! "module_binding_before_equal")
   (module_name) @append_indent_start @begin_scope
-  "="? @do_nothing
   [
-    (signature)? @do_nothing
     [
       (functor_type)
       (module_type_constraint)
     ] @append_indent_end @end_scope
   ]
```
Comment (nbacquey) [inline languages/ocaml.scm:1401]: We can remove double brackets here

**Comment by Xophmeister** [comment]
Thanks, @nbacquey. That updated query certainly worked :+1: and I put in the scoped softline in 37c5b47; thus un-drafting this PR. Note that I could only prepend the scoped softline (appending didn't do anything), so the final output looks like this:

```ocaml
module MFloat: SERIALISABLE with
  type t = float
= struct
  type t = float [@@deriving yojson]
  let _key = "float"
end
```

File: languages/ocaml.scm (hunk 1 above)
Comment (Xophmeister) [reply languages/ocaml.scm:1401]: Done in 37c5b47 :ok_hand:

**Comment by nbacquey** [comment]
Prepending is fine with me. Appending doesn't work because the custom scope ends just after the `=`, so the scoped softline would be out of scope (which should trigger a warning message).

**Pull request merged**